ecl/hqlcpp/hqlttcpp.cpp

```diff
@@ -8865,6 +8865,228 @@ IHqlExpression * FilteredIndexOptimizer::createTransformed(IHqlExpression * expr
+DFSLayoutTransformer::DFSLayoutTransformer(IErrorReceiver &_errs, ICodegenContextCallback * _ctxCallback, HqlCppOptions const &_options)
+: NewHqlTransformer(DFSLayoutTransformerInfo), errs(_errs), options(_options)
+{
+    ctxCallback = _ctxCallback;
+}
+
+
+IHqlExpression * DFSLayoutTransformer::createTransformed(IHqlExpression * expr)
+{
+    OwnedHqlExpr transformed = NewHqlTransformer::createTransformed(expr);
+    int filenameIdx = 0;
+    int recordIdx = 0;
+    int transformIdx = -1;
+    switch (transformed->getOperator())
+    {
+    default:
+        return transformed.getClear();
+    case no_table:
+        filenameIdx = 0;
+        recordIdx = 1;
+        break;
+    case no_keyindex:
+        throwUnexpected();  // They have already been translated into no_newkeyindex by HqlTreeNormalizer::transformKeyIndex
+        break;
+    case no_newkeyindex:
+        // Args are flatfile, record, transform, name. Offset of first payload field indicated by value of __payload_Atom attribute
+        filenameIdx = 3;
+        transformIdx = 2;
+        recordIdx = 1;
+        break;
+    }
+    LinkedHqlExpr ds = transformed;
+    IHqlExpression * filename = ds->queryChild(filenameIdx);
+    bool translate = getBoolAttribute(ds, lookupAtom, options.translateDFSlayouts && filename->isConstant());
+    if (translate)
+    {
+        ECLlocation where(ds);
+        if (!filename->isConstant())
+            errs.reportError(ERR_EXPECTED_CONST, "LOOKUP attribute requires a constant filename", str(where.sourcePath), where.lineno, where.column, where.position);
+        StringBuffer fileName;
+        filename->queryValue()->getStringValue(fileName);
+        if (fileName.length())   // PIPE creates a no_table with a blank filename, it seems
+        {
+            // Need to work out what to do about numKeyedFields and what happens if it changes
+            // If payload fields are added/removed, no problem, just transform as with datasets
+            // If keyed fields are added/removed, may be trickier
+            OwnedHqlExpr dfsLayout = ctxCallback->lookupDFSlayout(fileName, errs, where, ds->hasAttribute(optAtom));
+            if (dfsLayout)
+            {
+                IHqlExpression *wanted = ds->queryRecord();
+                // If wanted contains any virtual fields, clone them onto dfsLayout
+                if (containsVirtualFields(wanted))
+                {
+                    HqlExprArray newFields;
+                    unwindChildren(newFields, dfsLayout);
+                    getVirtualFields(wanted, newFields);
+                    dfsLayout.setown(dfsLayout->clone(newFields));
+                }
+                IHqlExpression *payload = dfsLayout->queryAttribute(_payload_Atom);
+                unsigned dfsPayload = 0;
+                if (payload)
+                {
+                    dfsPayload = payload->queryChild(0)->queryValue()->getIntValue();
```

Comment:
simpler: getInValue(payload->queryChild(0))

or even

getIntValue(queryAttributeChild(dfsLayout, _payload_Atom, 0)); 

if you were not modifying the dfsLayout.